File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveFieldAndPropertyAccess.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Transforms.InferTypesPlugins.Helpers.GenericsResolver as genRes

class ResolveFieldAndPropertyAccess(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveFieldAndPropertyAccess")
    
    def get_static_ref(self, cls_, member_name):
        field = next(filter(lambda x: x.name == member_name, cls_.fields), None)
        if field != None and field.is_static:
            return refs.StaticFieldReference(field)
        
        prop = next(filter(lambda x: x.name == member_name, cls_.properties), None)
        if prop != None and prop.is_static:
            return refs.StaticPropertyReference(prop)
        
        self.error_man.throw(f'''Could not resolve static member access of a class: {cls_.name}::{member_name}''')
        return None
    
    def get_instance_ref(self, cls_, member_name, obj):
        while True:
            field = next(filter(lambda x: x.name == member_name, cls_.fields), None)
            if field != None and not field.is_static:
                return refs.InstanceFieldReference(obj, field)
            
            prop = next(filter(lambda x: x.name == member_name, cls_.properties), None)
            if prop != None and not prop.is_static:
                return refs.InstancePropertyReference(obj, prop)
            
            if cls_.base_class == None:
                break
            
            cls_ = (cls_.base_class).decl
        
        self.error_man.throw(f'''Could not resolve instance member access of a class: {cls_.name}::{member_name}''')
        return None
    
    def get_interface_ref(self, intf, member_name, obj):
        field = next(filter(lambda x: x.name == member_name, intf.fields), None)
        if field != None and not field.is_static:
            return refs.InstanceFieldReference(obj, field)
        
        for base_intf in intf.base_interfaces:
            res = self.get_interface_ref((base_intf).decl, member_name, obj)
            if res != None:
                return res
        return None
    
    def transform_pa(self, expr):
        if isinstance(expr.object, refs.ClassReference):
            return self.get_static_ref(expr.object.decl, expr.property_name)
        
        if isinstance(expr.object, refs.StaticThisReference):
            return self.get_static_ref(expr.object.cls_, expr.property_name)
        
        expr.object = self.main.run_plugins_on(expr.object)
        
        if isinstance(expr.object, refs.ThisReference):
            return self.get_instance_ref(expr.object.cls_, expr.property_name, expr.object)
        
        type = expr.object.get_type()
        if isinstance(type, astTypes.ClassType):
            return self.get_instance_ref(type.decl, expr.property_name, expr.object)
        elif isinstance(type, astTypes.InterfaceType):
            ref = self.get_interface_ref(type.decl, expr.property_name, expr.object)
            if ref == None:
                self.error_man.throw(f'''Could not resolve instance member access of a interface: {type.repr()}::{expr.property_name}''')
            return ref
        elif type == None:
            self.error_man.throw(f'''Type was not inferred yet (prop="{expr.property_name}")''')
        elif isinstance(type, astTypes.AnyType):
            #this.errorMan.throw(`Object has any type (prop="${expr.propertyName}")`);
            expr.set_actual_type(astTypes.AnyType.instance)
        else:
            self.error_man.throw(f'''Expected class as variable type, but got: {type.repr()} (prop="{expr.property_name}")''')
        
        return expr
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.PropertyAccessExpression) and not (isinstance(expr.object, refs.EnumReference)) and not (isinstance(expr.parent_node, exprs.UnresolvedCallExpression) and expr.parent_node.func == expr) and not (isinstance(expr.actual_type, astTypes.AnyType))
    
    def transform(self, expr):
        return self.transform_pa(expr)
    
    def can_detect_type(self, expr):
        return isinstance(expr, refs.InstanceFieldReference) or isinstance(expr, refs.InstancePropertyReference) or isinstance(expr, refs.StaticFieldReference) or isinstance(expr, refs.StaticPropertyReference)
    
    def detect_type(self, expr):
        if isinstance(expr, refs.InstanceFieldReference):
            actual_type = genRes.GenericsResolver.from_object(expr.object).resolve_type(expr.field.type, True)
            expr.set_actual_type(actual_type, False, astTypes.TypeHelper.is_generic(expr.object.actual_type))
            return True
        elif isinstance(expr, refs.InstancePropertyReference):
            actual_type = genRes.GenericsResolver.from_object(expr.object).resolve_type(expr.property.type, True)
            expr.set_actual_type(actual_type)
            return True
        elif isinstance(expr, refs.StaticPropertyReference):
            expr.set_actual_type(expr.decl.type, False, False)
            return True
        elif isinstance(expr, refs.StaticFieldReference):
            expr.set_actual_type(expr.decl.type, False, False)
            return True
        
        return False

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveFieldAndPropertyAccess.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveFieldAndPropertyAccess;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\One\Ast\References\ClassReference;
use OneLang\One\Ast\References\Reference;
use OneLang\One\Ast\References\StaticFieldReference;
use OneLang\One\Ast\References\StaticPropertyReference;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\InstancePropertyReference;
use OneLang\One\Ast\References\ThisReference;
use OneLang\One\Ast\References\SuperReference;
use OneLang\One\Ast\References\EnumReference;
use OneLang\One\Ast\References\StaticThisReference;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Transforms\InferTypesPlugins\Helpers\GenericsResolver\GenericsResolver;

class ResolveFieldAndPropertyAccess extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveFieldAndPropertyAccess");
        
    }
    
    protected function getStaticRef($cls, $memberName) {
        $field = \OneCore\ArrayHelper::find($cls->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($field !== null && $field->isStatic)
            return new StaticFieldReference($field);
        
        $prop = \OneCore\ArrayHelper::find($cls->properties, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($prop !== null && $prop->isStatic)
            return new StaticPropertyReference($prop);
        
        $this->errorMan->throw("Could not resolve static member access of a class: " . $cls->name . "::" . $memberName);
        return null;
    }
    
    protected function getInstanceRef($cls, $memberName, $obj) {
        while (true) {
            $field = \OneCore\ArrayHelper::find($cls->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
            if ($field !== null && !$field->isStatic)
                return new InstanceFieldReference($obj, $field);
            
            $prop = \OneCore\ArrayHelper::find($cls->properties, function ($x) use ($memberName) { return $x->name === $memberName; });
            if ($prop !== null && !$prop->isStatic)
                return new InstancePropertyReference($obj, $prop);
            
            if ($cls->baseClass === null)
                break;
            
            $cls = ($cls->baseClass)->decl;
        }
        
        $this->errorMan->throw("Could not resolve instance member access of a class: " . $cls->name . "::" . $memberName);
        return null;
    }
    
    protected function getInterfaceRef($intf, $memberName, $obj) {
        $field = \OneCore\ArrayHelper::find($intf->fields, function ($x) use ($memberName) { return $x->name === $memberName; });
        if ($field !== null && !$field->isStatic)
            return new InstanceFieldReference($obj, $field);
        
        foreach ($intf->baseInterfaces as $baseIntf) {
            $res = $this->getInterfaceRef(($baseIntf)->decl, $memberName, $obj);
            if ($res !== null)
                return $res;
        }
        return null;
    }
    
    protected function transformPA($expr) {
        if ($expr->object instanceof ClassReference)
            return $this->getStaticRef($expr->object->decl, $expr->propertyName);
        
        if ($expr->object instanceof StaticThisReference)
            return $this->getStaticRef($expr->object->cls, $expr->propertyName);
        
        $expr->object = $this->main->runPluginsOn($expr->object);
        
        if ($expr->object instanceof ThisReference)
            return $this->getInstanceRef($expr->object->cls, $expr->propertyName, $expr->object);
        
        $type = $expr->object->getType();
        if ($type instanceof ClassType)
            return $this->getInstanceRef($type->decl, $expr->propertyName, $expr->object);
        else if ($type instanceof InterfaceType) {
            $ref = $this->getInterfaceRef($type->decl, $expr->propertyName, $expr->object);
            if ($ref === null)
                $this->errorMan->throw("Could not resolve instance member access of a interface: " . $type->repr() . "::" . $expr->propertyName);
            return $ref;
        }
        else if ($type === null)
            $this->errorMan->throw("Type was not inferred yet (prop=\"" . $expr->propertyName . "\")");
        else if ($type instanceof AnyType)
            //this.errorMan.throw(`Object has any type (prop="${expr.propertyName}")`);
            $expr->setActualType(AnyType::$instance);
        else
            $this->errorMan->throw("Expected class as variable type, but got: " . $type->repr() . " (prop=\"" . $expr->propertyName . "\")");
        
        return $expr;
    }
    
    function canTransform($expr) {
        return $expr instanceof PropertyAccessExpression && !($expr->object instanceof EnumReference) && !($expr->parentNode instanceof UnresolvedCallExpression && $expr->parentNode->func === $expr) && !($expr->actualType instanceof AnyType);
    }
    
    function transform($expr) {
        return $this->transformPA($expr);
    }
    
    function canDetectType($expr) {
        return $expr instanceof InstanceFieldReference || $expr instanceof InstancePropertyReference || $expr instanceof StaticFieldReference || $expr instanceof StaticPropertyReference;
    }
    
    function detectType($expr) {
        if ($expr instanceof InstanceFieldReference) {
            $actualType = GenericsResolver::fromObject($expr->object)->resolveType($expr->field->type, true);
            $expr->setActualType($actualType, false, TypeHelper::isGeneric($expr->object->actualType));
            return true;
        }
        else if ($expr instanceof InstancePropertyReference) {
            $actualType = GenericsResolver::fromObject($expr->object)->resolveType($expr->property->type, true);
            $expr->setActualType($actualType);
            return true;
        }
        else if ($expr instanceof StaticPropertyReference) {
            $expr->setActualType($expr->decl->type, false, false);
            return true;
        }
        else if ($expr instanceof StaticFieldReference) {
            $expr->setActualType($expr->decl->type, false, false);
            return true;
        }
        
        return false;
    }
}
